compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -0,0 +1,207 @@
+
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtOps.h"
+#include "iree/compiler/Codegen/Common/GPU/GPUPatterns.h"
+#include "iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.h"
+#include "iree/compiler/Codegen/Common/VectorLayoutAnalysis.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Utils.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/Verifier.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+
+namespace mlir::iree_compiler {
+
+using namespace mlir::iree_compiler::IREE::VectorExt;
+using VectorValue = TypedValue<VectorType>;
+
+// Helper to linearize the given |ids| with the given maximum values as |sizes|.
+// Gets the element ID in terms of |elementCount| and adds the element |offset|.
+static Value linearizeIndex(OpBuilder &builder, Value offset,
+                            ArrayRef<OpFoldResult> ids, ArrayRef<int64_t> sizes,
+                            int64_t elementCount) {
+  SmallVector<AffineExpr> exprs(ids.size() + 1);
+  bindSymbolsList(builder.getContext(), MutableArrayRef{exprs});
+  AffineExpr idExpr =
+      sizes[0] > 1 ? exprs[0] : builder.getAffineConstantExpr(0);
+
+  for (int i = 1, e = ids.size(); i < e; ++i) {
+    if (sizes[i] > 1) {
+      idExpr = idExpr * builder.getAffineConstantExpr(sizes[i]) + exprs[i];
+    }
+  }
+  idExpr = idExpr * builder.getAffineConstantExpr(elementCount);
+  idExpr = idExpr + exprs.back();
+  SmallVector<OpFoldResult> mapArgs(ids);
+  mapArgs.push_back(offset);
+  return affine::makeComposedAffineApply(
+             builder, offset.getLoc(),
+             AffineMap::get(0, mapArgs.size(), idExpr), mapArgs)
+      .getResult();
+}
+
+namespace {
+
+// Pattern to distribute `vector.transfer_read` ops with nested layouts.
+struct DistributeTransferReadNestedLayoutAttr final
+    : OpDistributionPattern<vector::TransferReadOp> {
+  using OpDistributionPattern::OpDistributionPattern;
+
+  DistributeTransferReadNestedLayoutAttr(MLIRContext *context, Value threadId)
+      : OpDistributionPattern(context), threadId(threadId) {}
+
+  LogicalResult matchAndRewrite(vector::TransferReadOp readOp,
+                                DistributionSignature &signature,
+                                PatternRewriter &rewriter) const override {
+    // TODO: Support masking.
+    if (readOp.getMask()) {
+      return failure();
+    }
+    NestedLayoutAttr vectorLayout =
+        dyn_cast<NestedLayoutAttr>(signature[readOp.getResult()]);
+    if (!vectorLayout) {
+      return failure();
+    }
+
+    if (!isa<MemRefType>(readOp.getSource().getType())) {
+      return failure();
+    }
```

Comment:
This pattern should work on a tensor also right?

Reply:
I'm not planning to add tests or holistic support for tensors so best to guard on unexpected usage.

Reply:
Makes sense